ironfish-cli/src/commands/accounts/repair.ts

```diff
@@ -0,0 +1,276 @@
+    {
+      name: 'account',
+      parse: (input: string): Promise<string> => Promise.resolve(input.trim()),
+      required: false,
+      description: 'Name of the account to repair the database for',
+    },
+  ]
+
+  async start(): Promise<void> {
+    const { args } = await this.parse(Repair)
+
+    const node = await this.sdk.node()
+    await NodeUtils.waitForOpen(node)
+
+    const account = this.loadAccount(node.wallet, args.account)
+
+    this.log(`Repairing wallet for account ${account.name}`)
+
+    this.log('Repairing decryptedNotes and balance')
+    await this.repairDecryptedNotes(account, node.wallet.walletDb, node.chain)
+
+    this.log('Repairing nullifierToNote')
+    await this.repairNullifierToNoteHash(account, node.wallet.walletDb)
+
+    this.log('Repairing sequenceToNoteHash')
+    await this.repairSequenceToNoteHash(account, node.wallet.walletDb, node.chain)
+  }
+
+  private loadAccount(wallet: Wallet, accountName: string | undefined): Account {
+    if (accountName) {
+      const account = wallet.getAccountByName(accountName)
+      if (account) {
+        return account
+      }
+      throw new Error(`No account with name ${accountName}`)
+    }
+
+    const defaultAccount = wallet.getDefaultAccount()
+    if (defaultAccount) {
+      return defaultAccount
+    }
+
+    throw new Error('Could not find an account to repair.')
+  }
+
+  private async repairDecryptedNotes(
+    account: Account,
+    walletDb: WalletDB,
+    chain: Blockchain,
+  ): Promise<void> {
+    let unconfirmedBalance = 0n
+
+    let noteUnspentMismatches = 0
+    let nullifierNoteHashMismatches = 0
+
+    await walletDb.db.transaction(async (tx) => {
+      for await (const decryptedNoteValue of account.getNotes()) {
+        const transactionValue = await account.getTransaction(
+          decryptedNoteValue.transactionHash,
+          tx,
+        )
+
+        Assert.isNotUndefined(
+          transactionValue,
+          'Account has a note but is missing the transaction that it received the note from. Account must be rescanned using `accounts:rescan --reset`',
+        )
+
+        await this.verifyBlockContainsTransaction(
+          decryptedNoteValue.transactionHash,
+          transactionValue.blockHash,
+          chain,
+        )
+
+        await walletDb.setNoteHashSequence(
+          account,
+          decryptedNoteValue.hash,
+          transactionValue.sequence,
+          tx,
+        )
+
+        if (!decryptedNoteValue.nullifier) {
+          if (transactionValue.sequence) {
+            throw new Error(
+              'Transaction marked as on chain, but note missing nullifier. Account must be rescanned using `accounts:rescan --reset`.',
+            )
+          }
+
+          continue
+        }
+
+        const spent = await chain.nullifiers.contains(decryptedNoteValue.nullifier)
+
+        if (spent && !decryptedNoteValue.spent) {
+          noteUnspentMismatches++
+
+          await walletDb.saveDecryptedNote(
```

Comment:
I feel like we could combine some of these operations with walletDB.updateDecryptedNote. Or if we don't trust that function create a similar one. I think we should update `decryptedNote.nullifier`, `decryptedNote.index`, `decryptedNote.spent` and `setNoteHashSequence` all at the same time since they should all be in sync

Reply:
Updated this method to only ever change notes from unspent to spent. The indexes won't change from that update, so it's more efficient (and maybe safer) to save the note directly.